### PR TITLE
Improve sourceMappingURL matching to prevent false-positives

### LINKED ISSUE #2408: Parcel with source maps enabled cannot import LiveScript
# 🐛 bug report

Parcel with source maps enabled cannot import LiveScript

## 🎛 Configuration (.babelrc, package.json, cli command)

```json
{
  "name": "livescript-parcel",
  "version": "1.0.0",
  "description": "",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "start": "parcel build index.js"
  },
  "keywords": [],
  "author": "",
  "license": "ISC",
  "dependencies": {
    "livescript": "^1.6.0",
    "parcel-bundler": "^1.10.3"
  }
}
```

## 🤔 Expected Behavior

`import livescript from 'livescript'` should work

## 😯 Current Behavior

```
 > parcel build index.js

⚠️  Could not load existing sourcemap of "node_modules/livescript/lib/index.js".
🚨  livescript-parcel/node_modules/livescript/lib/index.js:69:43: Expecting Unicode escape sequence \uXXXX (69:43)
  67 |         if ((ref$ = options.map) === 'linked' || ref$ === 'debug') {
  68 |           mapPath = path.basename(outputFilename) + ".map";
> 69 |           result.code += "\n + mapPath + "\n";
     |                                           ^
  70 |         } else {
  71 |           result.code += "\n//# sourceMappingURL=data:application/json;base64," + bufferFrom(result.map.toString()).toString('base64') + "\n";
  72 |         }
```

It appears that parcel is incorrectly interpreting the file as a source map.

```
        if ((ref$ = options.map) === 'linked' || ref$ === 'debug') {
          mapPath = path.basename(outputFilename) + ".map";
          result.code += "\n//# sourceMappingURL=" + mapPath + "\n";
        } else {
          result.code += "\n//# sourceMappingURL=data:application/json;base64," + bufferFrom(result.map.toString()).toString('base64') + "\n";
        }
```

## 💻 Code Sample

https://github.com/raine/livescript-parcel

### PR DESCRIPTION
**Pull request by lustoykov**
# ↪️ Pull Request
Fixes #2408 

## 💻 Examples

See the referenced issue (#2408) for an example by the author

## 🚨 Test instructions

Problem was a loose regex match that tried to find out if sourceMappingURL was referenced so parcel can load the sourcemaps. After changes in this PR have been made, I've written tests for the new regex match.

## ✔️ PR Todo

- [x] Added/updated unit tests for this change
- [x] Filled out test instructions (In case there aren't any unit tests)
- [x] Included links to related issues/PRs

Brainstormed a bit with @mischnic and we came up with this solution: sourceMappingURL is matched only if it's at the end of the bundle. This prevents false-positives where arbitrary variables named sourceMappingURL are included in the code. See #2408 for more context.

Seems like I moved too much stuff around for this fix, feedback appreciated.

### CODE REVIEW

File: packages/core/parcel-bundler/src/assets/JSAsset.js [hunk 1]
```diff
@@ -82,9 +85,14 @@ class JSAsset extends Asset {
     walk.ancestor(this.ast, collectDependencies, this);
   }
 
+  matchSourceMappingURL() {
```
Comment (DeMoorJasper) [inline packages/core/parcel-bundler/src/assets/JSAsset.js]: Why did this get moved into a seperate function?

Comment (lustoykov) [reply packages/core/parcel-bundler/src/assets/JSAsset.js]: only for testing purposes, I didn't want to test against `this.contents` directly, but rather against an (internal) API that `loadSourceMap` uses. I thought this way, if someone decides to adjust `matchSourceMappingURL`, the test would still kick in.

File: packages/core/parcel-bundler/test/assets/js-asset.js [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    const a = new JSAsset(__filename, {rootDir: '/root/dir'});
+    Object.assign(a, {
+      type: 'type',
+      contents: codeExample
+    });
+
+    assert(!a.matchSourceMappingURL());
+  });
+
+  it('should match sourceMappingURL when correctly inserted at end of the bundle', async () => {
```
Comment (DeMoorJasper) [inline packages/core/parcel-bundler/test/assets/js-asset.js]: Could you add a test for inline sourcemaps?

Comment (lustoykov) [reply packages/core/parcel-bundler/test/assets/js-asset.js]: oh, good catch. I've added another test case. I also wonder if there are other cases we should cover...

**Comment by devongovett** [comment]
Looks like the code moved to https://github.com/parcel-bundler/parcel/blob/master/packages/core/parcel-bundler/src/utils/loadSourceMap.js. Could you update?

File: packages/core/parcel-bundler/src/utils/loadSourceMap.js [hunk 3]
```diff
@@ -2,12 +2,16 @@ const logger = require('@parcel/logger');
 const path = require('path');
 const fs = require('@parcel/fs');
 
-const SOURCEMAP_RE = /(?:\/\*|\/\/)\s*[@#]\s*sourceMappingURL\s*=\s*([^\s*]+)(?:\s*\*\/)?/;
+const SOURCEMAP_RE = /(?:\/\*|\/\/)\s*[@#]\s*sourceMappingURL\s*=\s*([^\s*]+)(?:\s*\*\/)?\s*$/;
```
Comment (lustoykov) [inline packages/core/parcel-bundler/src/utils/loadSourceMap.js]: @mischnic I assume last changes on this regex were by you to make the css source mapping work. This change would only identify source maps at the end of the file. Hope this doesn't interfere with your changes?

**Comment by lustoykov** [comment]
@devongovett updated, `(Linux node_6_x)` fails, can't identify if related to this commit since tests on master fail too (with similar error?)

```
span>@parcel&#x2F;integration-tests: ‌success‌ Saved 1 new dependency.‌
@parcel&#x2F;integration-tests: ‌info‌ Direct dependencies‌
@parcel/integration-tests: └─ hyperapp@1.2.9
@parcel&#x2F;integration-tests: ‌info‌ All dependencies‌
@parcel/integration-tests: └─ hyperapp@1.2.9
@parcel/integration-tests: [9:59:29 AM]: Request: http://localhost:41323/
@parcel/integration-tests: error Command failed with exit code 1.
@parcel/integration-tests: info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
@parcel/integration-tests: npm ERR! Linux 4.15.0-1037-azure
@parcel/integration-tests: npm ERR! argv "/opt/hostedtoolcache/node/6.17.0/x64/bin/node" "/opt/hostedtoolcache/node/6.17.0/x64/bin/npm" "run" "test-ci"
@parcel/integration-tests: npm ERR! node v6.17.0
@parcel/integration-tests: npm ERR! npm  v3.10.10
@parcel/integration-tests: npm ERR! code ELIFECYCLE
@parcel/integration-tests: npm ERR! @parcel/integration-tests@1.11.0 test-ci: `yarn test --reporter mocha-multi-reporters --reporter-options configFile=./test/mochareporters.json`
@parcel/integration-tests: npm ERR! Exit status 1
```

**Comment by mischnic** [comment]
Doesn't seem related (you have to look in the *Tests* tab):

bildschirmfoto 2019-03-07 um 11 40 52

**Comment by lustoykov** [comment]
Thanks, I've been looking for this for a long time 🥇

File: packages/core/parcel-bundler/src/utils/loadSourceMap.js (hunk 3 above)
Comment (mischnic) [reply packages/core/parcel-bundler/src/utils/loadSourceMap.js]: my change also added the `*/` comment end tag, matching whitespace after that is correct.

**Comment by thscott** [comment]
Is there anything stopping this from being merged?

**Comment by walston** [comment]
Is there any way to get this merged into latest?
this would resolve an issue causing a breakage with `postcss` [see map-generator.es6#L142](https://github.com/postcss/postcss/blob/master/lib/map-generator.es6#L142).

**Comment by walston** [comment]
while this is waiting to be merged:
You can side-step the problem of false positives with `parcel --no-source-maps ...`

**Comment by AG89z** [comment]
Hi, when is this going to be merged?

**Comment by mjgerace** [comment]
Any update on the progress on this guy? It's affecting me.

**Comment by var** [comment]
This has been long pending. What might be holding it back? Anyway we can contribute? 

CC: @devongovett

**Comment by kurtharriger** [comment]
I just ran into this issue as well attempting to try the react-google-oauth library.  
```
⚠️  Could not load existing sourcemap of "node_modules/react-google-oauth/dist/react-google-oauth.js".
🚨  /Users/kharriger/pp/spendapps/daily-spending/node_modules/react-google-oauth/dist/react-google-oauth.js:6232:83: Unterminated regular expression (6232:83)
  6230 | 	if (sourceMap) {
  6231 | 		// http://stackoverflow.com/a/26603875
> 6232 | 		css += "\n + btoa(unescape(encodeURIComponent(JSON.stringify(sourceMap)))) + " */";
       | 		                                                                                ^
  6233 | 	}
  6234 |
  6235 | 	var blob = new Blob([css], { type: "text/css" });
⚠️  Could not load existing sourcemap of "node_modules/react-google-oauth/dist/react-google-oauth.js".
```

**Comment by mbohgard** [comment]
Why isn't this being merged? When can we expect a release with this fix? Thanks!

**Comment by DeMoorJasper** [comment]
@mbohgard this is probably not gonna get merged into master as we’re focusing on v2

<sub>Sent with <a href="http://githawk.com">GitHawk</a></sub>

**Comment by thscott** [comment]
@DeMoorJasper can we take from that that it'll be fixed in v2 then?

**Comment by DeMoorJasper** [comment]
@thscott Sure, I'll try port over this PR to v2 today

**Comment by amilanov75** [comment]
Apologies if this isn't the right thread. I keep getting this error when I build using Parcel, however everything appears to be working fine despite the error. Is this related to the source mapping issue of this thread? is it something that is being looked into?

image

**Comment by DeMoorJasper** [comment]
@amilanov75 this is just a warning the file is probably not published in the npm package but mentioned in the sourcemap under sources...

**Comment by amilanov75** [comment]
ok, I don't really understand what you are saying as I don't get what it means to publish a file in the npm package.
i installed the component via an npm package, but I'm guessing what you might say is that you can create a npm package in different ways and this one was created, but the file was not published - again, I don't understand the significance.

On the sourcemap side, I searched my hdd for "sourcemap" and found sourcemap folders which contained source-map.js files. In fact there are many files of differing naming conventions that contain "source-map" in the file name and when I look at them, some are very small files with a few lines of code with export statements, whereas others are enormous files with thousands of lines of code, so I'm none the wiser.

Could you kindly refer me to some locations, that written in human readable english ;) where I can read up on both the topics that you mentioned?

I assume in the end, this is not really a problem other than that I will have see the pecky error message every time I compile my code... which seems to be quite a lot i.e. each time I re-open VSC I need to recompile my project before I put a watch on it, after which i don't see the error anymore.

**Comment by DeMoorJasper** [comment]
@amilanov75 pretty sure that was English.

If you don't know what a sourcemap is google it, it's a pretty essential part of transforming and minifying JavaScript.

https://developer.mozilla.org/en-US/docs/Tools/Debugger/How_to/Use_a_source_map#:~:text=A%20source%20map%20is%20a,generate%20the%20source%20map

An npm package as you installed is just a folder with files that gets sent to a server somewhere (probably Azure), this folder can contain a bunch of files, sometimes people don't publish source files and only publish the compiled binaries, if this is the case the sourcemap can not map to these sourcefiels because they don't exist hence the warning... If this doesn't explain it than I don't know what will

**Comment by amilanov75** [comment]
Ha, ha, cheeky!!

I'm a noob to JS, hence all the questions.

Thank you for taking the time to explain.

**Pull request closed without merging**